### ISSUE TITLE
Constarints should be edge based

### ISSUE THREAD

**Issue by blaggacao** [odoo/odoo#34013]
Constraints should have access to both, previous and actual state.
Any edge based constraint now has to be written into the write method. Still it's a constraint, semantically.

**Comment by blaggacao** [comment]
@rco-odoo Maybe you can understand me... :wink:

**Comment by blaggacao** [comment]
https://speakerdeck.com/thockin/edge-vs-level-triggered-logic?slide=5

**Comment by blaggacao** [comment]
At the hart of it's inner self a validation logic is a transition logic (edge logic), or am I wrong?

**Comment by gdgellatly** [comment]
see https://docs.google.com/presentation/d/e/2PACX-1vR9VGsSQUnITdnQq5KRiWnY7o-yERCB8YfnqxzhFW5tFcJ8AzMmIQ1CpLTdEgy5Sz9nx-yQlC0BhyGR/pub?start=false&loop=false&delayms=3000&slide=id.p

**Comment by blaggacao** [comment]
@gdgellatly Good point! As long as we could query the two involved states, we could at least infer the transition. So level based is probably the right implementation paradigm, but we need both states to properly reason about the edge, itself.

**Comment by agenterp** [comment]
@gdgellatly nice approach! One Problem in its current implementation, the .yml state is not dynamic per object. In our case the conditions changes by service type.

**Comment by gdgellatly** [comment]
But service type is just a property of some object. And surely that's accessible via a computed or related field? As for ui, ultimately it's just got to read and write a yml file. I think ui tools already exist. Tbh it's just a project i know solves a lot of similar problems, i have no involvement in it.

**Comment by pedrovagner** [comment]
I made an application that validates input values, groups, and domains before and after create and write. Everything can be edited via UI.

![image](https://user-images.githubusercontent.com/10199950/180071647-c7e51ed3-61e6-4b85-a80f-04bc9bc4f26f.png)

Should I make it open source?
